rust/src/dcerpc/dcerpc.rs

```diff
@@ -0,0 +1,2600 @@
+    pub call_id: u32,
+}
+
+#[derive(Debug)]
+pub struct DCERPCHdrUdp {
+    pub rpc_vers: u8,
+    pub pkt_type: u8,
+    pub flags1: u8,
+    pub flags2: u8,
+    pub drep: Vec<u8>,
+    pub serial_hi: u8,
+    pub objectuuid: Vec<u8>,
+    pub interfaceuuid: Vec<u8>,
+    pub activityuuid: Vec<u8>,
+    pub server_boot: u32,
+    pub if_vers: u32,
+    pub seqnum: u32,
+    pub opnum: u16,
+    pub ihint: u16,
+    pub ahint: u16,
+    pub fraglen: u16,
+    pub fragnum: u16,
+    pub auth_proto: u8,
+    pub serial_lo: u8,
+}
+
+#[derive(Debug)]
+pub struct DCERPCBind {
+    pub numctxitems: u8,
+    pub uuid_list: Vec<DCERPCUuidEntry>,
+}
+
+#[derive(Debug)]
+pub struct BindCtxItem {
+    pub ctxid: u16,
+    pub uuid: Vec<u8>,
+    pub version: u16,
+    pub versionminor: u16,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct DCERPCBindAckResult {
+    pub ack_result: u16,
+    pub ack_reason: u16,
+    pub transfer_syntax: Vec<u8>,
+    pub syntax_version: u32,
+}
+
+#[derive(Debug)]
+pub struct DCERPCBindAck {
+    pub accepted_uuid_list: Vec<DCERPCUuidEntry>,
+    pub sec_addr_len: u16,
+    pub numctxitems: u8,
+    pub ctxitems: Vec<DCERPCBindAckResult>,
+}
+
+#[derive(Debug)]
+pub struct DCERPCState {
+    pub header: Option<DCERPCHdr>,
+    pub bind: Option<DCERPCBind>,
+    pub bindack: Option<DCERPCBindAck>,
+    pub request: Option<DCERPCRequest>,
+    pub response: Option<DCERPCResponse>,
+    pub buffer_ts: Vec<u8>,
+    pub buffer_tc: Vec<u8>,
+    pub pad: u8,
+    pub padleft: u16,
+    pub bytes_consumed: u16,
+    pub tx_id: u16,
+    pub query_completed: bool,
+    pub data_needed_for_dir: u8,
+    pub prev_dir: u8,
+    pub detect_flags_ts: u64,
+    pub detect_flags_tc: u64,
+    pub de_state: Option<*mut core::DetectEngineState>,
+}
+
+impl DCERPCState {
+    pub fn new() -> DCERPCState {
+        return DCERPCState {
+            header: None,
+            bind: None,
+            bindack: None,
+            request: None,
+            response: None,
+            buffer_ts: Vec::new(),
+            buffer_tc: Vec::new(),
+            pad: 0,
+            padleft: 0,
+            bytes_consumed: 0,
+            tx_id: 0,
+            query_completed: false,
+            data_needed_for_dir: core::STREAM_TOSERVER,
+            prev_dir: core::STREAM_TOSERVER,
+            detect_flags_ts: 0,
+            detect_flags_tc: 0,
+            de_state: None,
+        };
+    }
+
+    fn get_hdr_drep_0(&self) -> u8 {
+        if let Some(ref hdr) = &self.header {
+            return hdr.packed_drep[0];
+        }
+        0
+    }
+
+    fn get_endianness(&self) -> Endianness {
+        let drep_0 = self.get_hdr_drep_0();
+        if drep_0 & 0x10 == 0 {
+            return Endianness::Big;
+        }
+        Endianness::Little
+    }
+
+    fn get_hdr_fraglen(&self) -> Option<u16> {
+        if let Some(ref hdr) = self.header {
+            return Some(hdr.frag_length);
+        }
+        // Shouldn't happen
```

Comment:
@catenacyber @jasonish it would be good if we can create a Rust equivalent of `DEBUG_VALIDATE_BUG_ON` for a case like this, so that the fuzzers can see if this is reachable.

Reply:
@inashivb maybe you can create it. It should be a wrapper around `panic!` that is only in effect if the `--enable-debug-validation` configure flag is set.

Reply:
Leaving it to you @inashivb :-)

Reply:
Hi @catenacyber and @victorjulien !
Are you looking for something like the following? Just want to confirm that I understood the requirement correctly.

```
+  17 #[macro_export]
+  18 macro_rules! validate_item (
+  19   ($item:expr) => {
+  20     if $item.is_none() {
+  21         panic!("BUG: {:?} should not be NULL!");
+  22     }
+  23   };
+  24 );
```
And then this function in `dcerpc.rs` would look like
```
~  317     fn get_hdr_fraglen(&self) -> u16 {
~  318         validate_item!(self.header);
~  319         self.header.as_ref().unwrap().frag_length
    320     }
```

Reply:
Hi @catenacyber and @victorjulien ! Is the following patch in the right direction? Just want to know if I understood the requirement correctly.
```
diff --git a/rust/Makefile.am b/rust/Makefile.am
index 158d348b7..b9d86ab75 100644
--- a/rust/Makefile.am
+++ b/rust/Makefile.am
@@ -19,6 +19,10 @@ if DEBUG
 RUST_FEATURES +=       debug
 endif
 
+if DEBUG_VALIDATION
+RUST_FEATURES +=>   debug-validate
+endif
+
 if RUST_CROSS_COMPILE 
 RUST_TARGET = --target $(host_triplet)
 endif
diff --git a/rust/src/common.rs b/rust/src/common.rs
index 668d55c86..0956e340e 100644
--- a/rust/src/common.rs
+++ b/rust/src/common.rs
@@ -14,6 +14,24 @@ macro_rules! take_until_and_consume (
   );
 );
 
+
+#[cfg(not(feature = "debug-validate"))]
+#[macro_export]
+macro_rules! validate_option_item (
+  ($item:expr) => {};
+);
+
+
+#[cfg(feature = "debug-validate")]
+#[macro_export]
+macro_rules! validate_option_item (
+  ($item:expr) => {
+    if $item.is_none() {
+        panic!("BUG: {:?} should not be NULL!");
+    }
+  };
```
and then in `dcerpc.rs`:
```
diff --git a/rust/src/dcerpc/dcerpc.rs b/rust/src/dcerpc/dcerpc.rs
index 7d9cbaf85..850966c53 100644
--- a/rust/src/dcerpc/dcerpc.rs
+++ b/rust/src/dcerpc/dcerpc.rs
@@ -315,6 +315,7 @@ impl DCERPCState {
     }
 
     fn get_hdr_fraglen(&self) -> Option<u16> {
+        validate_option_item!(self.header);
         if let Some(ref hdr) = self.header {
             return Some(hdr.frag_length);
         }
```

Reply:
I think the argument should be a bool. Think it would also be good to use the same name as in C.